app/models/experimentation/notifications_experiment.rb

```diff
@@ -90,6 +90,39 @@ def evict_patients
       cancel_evicted_notifications
     end
 
+    def mark_visits
+      treatment_group_memberships.status_enrolled.in_batches(of: 1000).each_record do |membership|
+        bp = BloodPressure
+          .where(patient_id: membership.patient_id)
+          .where("recorded_at > ?", membership.experiment_inclusion_date)
+          .select(:id, :recorded_at, :device_created_at, :facility_id)
+          .min_by(&:recorded_at)
+        bs = BloodSugar
+          .where(patient_id: membership.patient_id)
+          .where("recorded_at > ?", membership.experiment_inclusion_date)
+          .select(:id, :recorded_at, :device_created_at, :facility_id)
+          .min_by(&:recorded_at)
+        pd = PrescriptionDrug
+          .where(patient_id: membership.patient_id)
+          .where("device_created_at > ?", membership.experiment_inclusion_date)
+          .select(:id, :device_created_at, :facility_id)
+          .min_by(&:device_created_at)
+
+        earliest_visit = [bp, bs, pd].compact.min_by(&:device_created_at)
```

Comment:
How about moving this to `membership#earliest_visit`?

Reply:
Also, have we analysed performance? Is this good enough? We can consider running this `each_record` in parallel at the application level for an easy / simple performance boost.

Reply:
This isn't performant enough, unfortunately. It takes ~15ms per patient, which is good in isolation but since we'll be doing this for 100-200k patients everyday which is roughly 30-40 minutes. I've been trying to do this with SQL although that seems hard to get right. Will try parallelization if that doesn't work out 👍🏼 .